src/openrct2/ride/Vehicle.cpp

```diff
@@ -9890,3 +9899,150 @@ void vehicle_invalidate_window(rct_vehicle * vehicle)
 
     window_invalidate(w);
 }
+
+void vehicle_update_crossings(rct_vehicle * vehicle)
+{
+    if (vehicle_get_head(vehicle) != vehicle)
+    {
+        return;
+    }
+
+    rct_vehicle *frontVehicle;
+    rct_vehicle *backVehicle;
+
+    bool travellingForwards = !(vehicle->update_flags & VEHICLE_UPDATE_FLAG_3);
+
+    if (travellingForwards)
+    {
+        frontVehicle = vehicle;
+        backVehicle = vehicle_get_tail(vehicle);
+    }
+    else
+    {
+        frontVehicle = vehicle_get_tail(vehicle);
+        backVehicle = vehicle;
+    }
+
+    rct_xy_element xyElement;
+    track_begin_end output;
+    sint32 z, direction;
+
+    xyElement.x = frontVehicle->track_x;
+    xyElement.y = frontVehicle->track_y;
+    z           = frontVehicle->track_z;
+    xyElement.element = map_get_track_element_at_of_type_seq(
+        frontVehicle->track_x, frontVehicle->track_y, frontVehicle->track_z >> 3,
+        frontVehicle->track_type >> 2, 0
+        );
+    
+    if (xyElement.element && vehicle->status != VEHICLE_STATUS_ARRIVING)
+    {
+        sint16  autoReserveAhead = 4 + abs(vehicle->velocity) / 150000;
+        sint16  crossingBonus = 0;
+        bool    playedClaxon = false;
+
+            // vehicle positions mean we have to take larger
+            //  margins for travelling backwards
+        if (!travellingForwards)
+        {
+            autoReserveAhead += 1;
+        }
+
+        while (true)
+        {
+            rct_tile_element *tileElement = map_get_path_element_at(
+               xyElement.x / 32,
+               xyElement.y / 32,
+               xyElement.element->base_height
+            );
+
+            if (tileElement)
+            {
+                if (!playedClaxon && 0 == (tileElement->flags & TILE_ELEMENT_FLAG_BLOCKED_BY_VEHICLE))
+                {
+                    vehicle_claxon(vehicle);
```

Comment:
vehicle_claxon or is it a crossing claxon? I've not actually tested this so it might be a silly question.